src/Storages/StorageMerge.cpp

```diff
@@ -783,12 +849,119 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
+        /// Besides this we add FilterTransform if it is needed to follow row level policies.
+
+        convertingSourceStream(header,
+            storage_snapshot->metadata,
+            aliases,
+            modified_context,
+            *builder,
+            processed_stage,
+            row_policy_data);
     }
 
     return builder;
 }
 
+void ReadFromMerge::RowPolicyData::init(RowPolicyFilterPtr row_policy_filter_ptr_,
+    std::shared_ptr<DB::IStorage> storage,
+    ContextPtr local_context)
+{
+    if (row_policy_filter_ptr_)
+    {
+        row_policy_filter_ptr = row_policy_filter_ptr_;
+
+        ASTPtr expr = row_policy_filter_ptr->expression;
+
+        auto storage_metadata_snapshot = storage->getInMemoryMetadataPtr();
+        auto storage_columns = storage_metadata_snapshot->getColumns();
+        auto needed_columns = storage_columns.getAllPhysical();
+
+        auto syntax_result = TreeRewriter(local_context).analyze(expr, needed_columns);
+        auto expression_analyzer = ExpressionAnalyzer{expr, syntax_result, local_context};
+
+        actions_dag = expression_analyzer.getActionsDAG(false /* add_aliases */, false /* project_result */);
+        filter_actions = std::make_shared<ExpressionActions>(actions_dag,
+            ExpressionActionsSettings::fromContext(local_context, CompileExpressions::yes));
+        filter_column_name = namesDifference(filter_actions->getSampleBlock().getNames(), filter_actions->getRequiredColumns());
+    }
+}
+
+// Add columns that needed to evaluate row policies
+// SELECT x from t  if  t has row policy
+void ReadFromMerge::RowPolicyData::extendNames(Names & names)
+{
+    assert(row_policy_filter_ptr);
+    ASTPtr expr = row_policy_filter_ptr->expression;
+
+    RequiredSourceColumnsVisitor::Data columns_context;
+    RequiredSourceColumnsVisitor(columns_context).visit(expr);
+
+    const auto req_columns = columns_context.requiredColumns();
+
+    std::sort(names.begin(), names.end());
+    NameSet added_names;
+
+    for (const auto & req_column : req_columns)
+    {
+        if (!std::binary_search(names.begin(), names.end(), req_column))
+        {
+            added_names.insert(req_column);
+        }
+    }
+    if (!added_names.empty())
+    {
+        std::copy(added_names.begin(), added_names.end(), std::back_inserter(names));
+    }
+}
+
+void ReadFromMerge::RowPolicyData::addStorageFilter(SourceStepWithFilter * step)
+{
+    assert(row_policy_filter_ptr);
+    LOG_TRACE(&Poco::Logger::get("ReadFromMerge::RowPolicyData::addStorageFilter"), "filter_actions_dag: {},<> {}, <> {}",
+        filter_actions->getActionsDAG().dumpNames(),
+        filter_actions->getActionsDAG().dumpDAG(),
+        filter_actions->getSampleBlock().dumpStructure());
+
+    step->addFilter(actions_dag, filter_column_name);
+}
+
+void ReadFromMerge::RowPolicyData::addFilterTransform(QueryPipelineBuilder & builder)
+{
+    assert(row_policy_filter_ptr);
```

Comment:
`std::optional<RowPolicyData>` (see my [other comment](https://github.com/ClickHouse/ClickHouse/pull/50209#discussion_r1358108401)) would allow to get rid of such asserts.